src/node/node_class.hpp

```diff
@@ -297,111 +1330,195 @@ template<typename ClassType>
-    v8::Isolate* isolate = info.GetIsolate();
-    node::ReturnValue return_value(info.GetReturnValue());
-    try {
-        F(isolate, info.This(), property, return_value);
-    }
-    catch (std::exception &e) {
-        Nan::ThrowError(node::Exception::value(isolate, e));
-    }
+Napi::Value wrap(const Napi::CallbackInfo& info, const Napi::Object& instance, const Napi::String& property) {
+	Napi::Env env = info.Env();
+	node::ReturnValue result(env);
+
+	try {
+		F(env, instance, property, result);
+		return result.ToValue();
+	}
+	catch (const Napi::Error& e) {
+		throw;
+	}
+	catch (const node::Exception& e) {
+		Napi::Error error = Napi::Error::New(info.Env(), e.what());
+		copyObject(env, e.m_value, error);
+		throw error;
+	}
+	catch (const std::exception& e) {
+		throw Napi::Error::New(info.Env(), e.what());
+	}
 }
 
 template<node::StringPropertyType::SetterType F>
-void wrap(v8::Local<v8::String> property, v8::Local<v8::Value> value, const v8::PropertyCallbackInfo<v8::Value>& info) {
-    v8::Isolate* isolate = info.GetIsolate();
-    try {
-        if (F(isolate, info.This(), property, value)) {
-            // Indicate that the property was intercepted.
-            info.GetReturnValue().Set(value);
-        }
-    }
-    catch (std::exception &e) {
-        Nan::ThrowError(node::Exception::value(isolate, e));
-    }
+Napi::Value wrap(const Napi::CallbackInfo& info, const Napi::Object& instance, const Napi::String& property, const Napi::Value& value) {
+	Napi::Env env = info.Env();
+	try {
+		bool success = F(env, instance, property, value);
+		return Napi::Value::From(env, success);
+	}
+	catch (const Napi::Error& e) {
+		throw;
+	}
+	catch (const node::Exception& e) {
+		Napi::Error error = Napi::Error::New(info.Env(), e.what());
+		copyObject(env, e.m_value, error);
+		throw error;
+	}
+	catch (const std::exception& e) {
+		throw Napi::Error::New(info.Env(), e.what());
+	}
 }
 
 template<node::StringPropertyType::EnumeratorType F>
-void wrap(const v8::PropertyCallbackInfo<v8::Array>& info) {
-    auto names = F(info.GetIsolate(), info.This());
-    int count = (int)names.size();
-    v8::Local<v8::Array> array = Nan::New<v8::Array>(count);
-
-    for (int i = 0; i < count; i++) {
-        Nan::Set(array, i, v8::Local<v8::String>(names[i]));
-    }
-
-    info.GetReturnValue().Set(array);
+Napi::Value wrap(const Napi::CallbackInfo& info, const Napi::Object& instance) {
+	Napi::Env env = info.Env();
+
+	try {
+		auto names = F(env, instance);
+
+		int count = (int)names.size();
+		Napi::Array array = Napi::Array::New(env, count);
+		for (int i = 0; i < count; i++) {
+			array.Set(i, names[i]);
+		}
+
+		return array;
+	}
+	catch (const Napi::Error& e) {
+		throw;
+	}
+	catch (const node::Exception& e) {
+		Napi::Error error = Napi::Error::New(info.Env(), e.what());
+		copyObject(env, e.m_value, error);
+		throw error;
+	}
+	catch (const std::exception& e) {
+		throw Napi::Error::New(info.Env(), e.what());
+	}
```

Comment:
I see this catch pattern reused many times in the code - would there be a way to move it to a shared location where it can be reused?

Reply:
yes it could. using a define I just don't like how readable that gets

```
try {
  F(env, instanceProxy, args, result);
  return result.ToValue();
}
HANDLE_WRAP_EXCEPTION

```

but I agree on the code duplication concern

Reply:
What about a method that takes a lambda instead? That would be how I would solve it in JavaScript.

Reply:
that won't work. There is way to do that in c++ but it will be even more unreadable and involves more code and will have a bit of performance penalty. I don't think it's worth it

Reply:
went with the define code above